libpod/container_graph.go

```diff
@@ -0,0 +1,174 @@
+package libpod
+
+import (
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type containerNode struct {
+	id         string
+	container  *Container
+	dependsOn  []*containerNode
+	dependedOn []*containerNode
+}
+
+type containerGraph struct {
+	nodes              map[string]*containerNode
+	noDepNodes         []*containerNode
+	notDependedOnNodes []*containerNode
+}
+
+func buildContainerGraph(ctrs []*Container) (*containerGraph, error) {
+	graph := new(containerGraph)
+	graph.nodes = make(map[string]*containerNode)
+
+	// Start by building all nodes, with no edges
+	for _, ctr := range ctrs {
+		ctrNode := new(containerNode)
+		ctrNode.id = ctr.ID()
+		ctrNode.container = ctr
+
+		graph.nodes[ctr.ID()] = ctrNode
+	}
+
+	// Now add edges based on dependencies
+	for _, node := range graph.nodes {
+		deps := node.container.Dependencies()
+		for _, dep := range deps {
+			// Get the dep's node
+			depNode, ok := graph.nodes[dep]
+			if !ok {
+				return nil, errors.Wrapf(ErrNoSuchCtr, "container %s depends on container %s not found in input list", node.id, dep)
+			}
+
+			// Add the dependent node to the node's dependencies
+			// And add the node to the dependent node's dependedOn
+			node.dependsOn = append(node.dependsOn, depNode)
+			depNode.dependedOn = append(depNode.dependedOn, node)
+		}
+
+		// Maintain a list of nodes with no dependencies
+		// (no edges coming from them)
+		if len(deps) == 0 {
+			graph.noDepNodes = append(graph.noDepNodes, node)
+		}
+	}
+
+	// Need one more loop to get nodes that have nothing depending on them
```

Comment:
Could you make notDependsOnNodes be a map and populate it with an entry near line 31, then set that value for that entry to nil if you get a hit on depenedOn near line 47?  I think you could avoid another loop that way and perhaps have faster lookups than an array.  Might have a very sparsely filled map however.

Reply:
Good call. Avoiding that last loop would be nice.

Reply:
Fixed